x/evm/types/state_transition.go

```diff
@@ -109,6 +109,28 @@ func (st StateTransition) newEVM(
 	return vm.NewEVM(blockCtx, txCtx, csdb, config.EthereumConfig(st.ChainID), vmConfig)
 }
 
+func (st StateTransition) refundGas(gasInfo GasInfo) {
+	// Apply refund counter, capped to half of the used gas.
```

Comment:
why is it capped to half?

Reply:
this is how it's implemented on geth and was just ripped from there https://github.com/ethereum/go-ethereum/blob/d8ff53dfb8a516f47db37dbc7fd7ad18a1e8a125/core/state_transition.go#L288